runtime/gc_base/IndexableObjectAllocationModel.cpp

```diff
@@ -287,5 +321,73 @@ MM_IndexableObjectAllocationModel::layoutDiscontiguousArraylet(MM_EnvironmentBas
 	return spine;
 }
 
+#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
+#if !defined(LINUX) || !defined(J9VM_ENV_DATA64) || defined(OMRZTPF)
+/* Double map is only supported on LINUX 64 bit Systems for now */
+#error "Platform not supported by Double Map API"
+#endif /* !definedLINUX || !defined(J9VM_ENV_DATA64) || defined(OMRZTPF) */
+void* 
+MM_IndexableObjectAllocationModel::doubleMapArraylets(MM_EnvironmentBase* env, J9Object *objectPtr) 
+{
+	MM_GCExtensions *extensions = MM_GCExtensions::getExtensions(env);
+	J9JavaVM *javaVM = extensions->getJavaVM();
+	PORT_ACCESS_FROM_ENVIRONMENT(env);
+
+	GC_ArrayletLeafIterator arrayletLeafIterator(javaVM, (J9IndexableObject*)objectPtr);
+	UDATA arrayletLeafCount = arrayletLeafIterator.getNumLeafs();
+	MM_Heap *heap = extensions->getHeap();
+
+	void *result = NULL;
+
+#define ARRAYLET_ALLOC_THRESHOLD 64
+	void *leaves[ARRAYLET_ALLOC_THRESHOLD];
+	void **arrayletLeaveAddrs = leaves;
+	if (arrayletLeafCount > ARRAYLET_ALLOC_THRESHOLD) {
+		arrayletLeaveAddrs = (void **)env->getForge()->allocate(arrayletLeafCount * sizeof(uintptr_t), MM_AllocationCategory::GC_HEAP, J9_GET_CALLSITE());
+	}
+
+	if(arrayletLeaveAddrs == NULL) {
+		return NULL;
+	}
+	UDATA elementsSize = extensions->indexableObjectModel.getDataSizeInBytes((J9IndexableObject*)objectPtr);
+
+	GC_SlotObject *slotObject = NULL;
+	uintptr_t count = 0;
 
+	while (NULL != (slotObject = arrayletLeafIterator.nextLeafPointer())) {
+		void *currentLeaf = slotObject->readReferenceFromSlot();
+		arrayletLeaveAddrs[count] = currentLeaf;
+		count++;
+	}
+
+	J9Object *firstLeafSlot = (J9Object *)((uintptr_t)extensions->indexableObjectModel.getArrayoidPointer((J9IndexableObject*)objectPtr)[0]);
+
```

Comment:
I believe this requires one more step, to decompress slot value, before going to look up for region descriptor